store/etcd/rulesets.go

```diff
@@ -164,14 +159,28 @@ func (s *RulesetService) listRulesets(ctx context.Context, key, prefix string, l
 		return nil, store.ErrNotFound
 	}
 
+	ops := make([]clientv3.Op, 0, resp.Count)
+	txn := s.Client.KV.Txn(ctx)
+	for _, pair := range resp.Kvs {
+		ops = append(ops, clientv3.OpGet(string(pair.Value)))
+	}
+	txnresp, err := txn.Then(ops...).Commit()
+	if err != nil {
+		return nil, errors.Wrap(err, "transaction failed to fetch all entries")
+	}
+
 	var entries store.RulesetEntries
 	entries.Revision = strconv.FormatInt(resp.Header.Revision, 10)
 	entries.Entries = make([]store.RulesetEntry, len(resp.Kvs))
-	for i, pair := range resp.Kvs {
-		err = json.Unmarshal(pair.Value, &entries.Entries[i])
-		if err != nil {
-			s.Logger.Debug().Err(err).Bytes("entry", pair.Value).Msg("list: unmarshalling failed")
-			return nil, errors.Wrap(err, "failed to unmarshal entry")
+
+	for i, resps := range txnresp.Responses {
+		rr := resps.GetResponseRange()
+		for _, pair := range rr.Kvs {
+			err = json.Unmarshal(pair.Value, &entries.Entries[i])
```

Comment:
I'm not sure but I think there is a mistake here, you are using the `i` of the outside loop, which means you are overwriting the same entry every time, aren't you?

Reply:
There is not. Because the `Responses` slice handles the responses for the whole transaction (related to the number of OpGet calls), for example if there is 3 calls to `OpGet` we will 3 entries in the slice.
Whereas the `Kvs` one handles the responses for each calls, and given that we are getting a leaf in the tree we will always have one value in the `Kvs` slice. 
I updated the code in https://github.com/heetch/regula/pull/59/commits/25e916a6b0821c48071a78ea9a8694be84e7c07a. Is it more clear?